### DISCUSSION TITLE
Can you update parameters of a module from inside the nn.compact of that module? (self modifying networks)

### DISCUSSION THREAD

**Discussion by MOVzeroOne** [google/flax#2208] [Q&A] [answered]
I'm quite new to flax and I was wondering what the correct way is to get this behavior: 
param = f.init(key,x)
new_param, y = f.apply(param,x)
Where f is a nn.module instance.
Where f might go through multiple operations to get new_param and that those operations might rely on the intermediate param to produce their output.

So basically, is there a way I can access and update the parameters supplied to an instance of  nn.module from within the ```__call__```,
while not losing the functional property so it can all be wrapped with the grad function transform.

**Comment by MOVzeroOne** [comment, marked as answer]
I found how to get the variables inside a module https://github.com/google/flax/discussions/1846
solution is https://github.com/google/flax/discussions/2215. (just notice that all variables tied to the original parameters dont get updated, grads are traced through self.put_variable())

**Reply by MOVzeroOne** [reply]
So I came up with this
```
import jax 
from jax import numpy as jnp 
from jax import grad,random,jit,vmap 
import flax 
from flax import linen as nn 



class network(nn.Module):
    input_size : int 
    output_size : int 
    @nn.compact
    def __call__(self,x):
        W = self.param('W',nn.initializers.normal(),(self.input_size,self.output_size))
        b = self.param('b',nn.initializers.normal(),(self.output_size,))

      
        self.put_variable("params","b",(x@W+b).reshape(5,))  
    
        return jnp.sum(x+b)


if __name__ == "__main__":
    key = random.PRNGKey(0)
    key_x,key_param,key = random.split(key,3)
    x = random.normal(key_x,(1,5))

    module = network(5,5)
    param = module.init(key_param,x)
    print(param)
    #x,param = module.apply(param,x,mutable=["params"])
    #print(param)
    print(grad(module.apply,has_aux=True)(param,x,mutable=["params"]))
```
problem now is that the grads arent traced through the self.put_variable method... :(

**Reply by jheek** [reply]
The more canonical way to do this would be: 
```
W = self.variable('params', 'W', lambda: nn.initializers.normal()(self.make_rng('params'), (self.input_size,self.output_size)))
W.value = x @ W.value + b).reshape(5,)
```

**Reply by MOVzeroOne** [reply]
Thank you :)
